[deckhouse] Add Webhook-operator

riptide-01 opened the pull request:
## Description

### 🎯 Webhook Operator

**Webhook-operator** is a Kubernetes operator that allows you to define and manage shell-operator webhooks as Kubernetes resources.

It provides a declarative interface for creating `ValidationWebhook` and `ConversionWebhook` Custom Resources - validation or conversion logic is defined directly in the manifest using Python code.

<details> <summary>🔄 <b>How It Works (Flow)</b></summary>

```
┌─────────────────────────┐
│  Create Webhook CR      │
└───────────┬─────────────┘
            ▼
┌─────────────────────────┐
│  Operator generates     │
│  Python hook file       │
└───────────┬─────────────┘
            ▼
┌─────────────────────────┐
│  Shell-operator reloads │
│  and registers webhook  │
└───────────┬─────────────┘
            ▼
┌─────────────────────────┐
│  Kubernetes calls       │
│  webhook on requests    │
└─────────────────────────┘
```

</details>

---
### 🗒️ Usage examples

#### 📋 Validation Webhook

Validates Kubernetes resources on CREATE/UPDATE/DELETE operations.

**Example:** Deny Services with `test` in the name:
```yaml
apiVersion: deckhouse.io/v1alpha1
kind: ValidationWebhook
metadata:
  name: service-name-validator
validationObject:
  name: service-validator.example.deckhouse.io
  rules:
  - apiGroups:   [""]
    apiVersions: ["v1"]
    operations:  ["CREATE", "UPDATE"]
    resources:   ["services"]
    scope:       "Namespaced"
  namespace:
    labelSelector:
      matchLabels:
        validation-enabled: "true"
  failurePolicy: Fail
  sideEffects: None
handler:
  python: |
    def validate(ctx: DotMap) -> tuple[Optional[str], bool]:
        name = ctx.review.request.object.metadata.name
        
        if "test" in name:
            return f"Service '{name}' contains forbidden word", False
        
        return None, True
```

**Handler return type:** `tuple[Optional[str], bool]`

| Result                | Return Value                |
| --------------------- | --------------------------- |
| ✅ Allow               | `(None, True)`              |
| ❌ Deny                | `("error message", False)`  |
| ⚠️ Allow with warning | `("warning message", True)` |
<details><summary><b>🎨 Interactive Example</b> (Full flow)</summary>
<a href="https://asciinema.org/a/rFcOS7jRKvgHbXoc">
  💾 asciinema player
</a>

<img alt="conversion_demo.gif" src="https://github.com/user-attachments/assets/1af9299a-7a23-4ea1-9654-df0ef31e2e12"/>
</details>

---
#### 🔀 Conversion Webhook

Converts CRD resources between API versions on-the-fly (for migrations).

**Example:** Convert between `v1beta1` and `v1`:

```yaml
apiVersion: deckhouse.io/v1alpha1
kind: ConversionWebhook
metadata:
  name: example.deckhouse.io
conversions:
  - from: v1beta1
    to: v1
    handler:
      python: |
        def v1beta1_to_v1(self, o: dict) -> typing.Tuple[str | None, dict]:
            obj = DotMap(o)
            obj.apiVersion = "deckhouse.io/v1"
            
            # Split hostPort into separate fields
            if "hostPort" in obj.spec:
                host, port = obj.spec.hostPort.rsplit(":", 1)
                obj.spec.host = host
                obj.spec.port = port
                del obj.spec.hostPort
            
            return None, obj.toDict()

  - from: v1
    to: v1beta1
    handler:
      python: |
        def v1_to_v1beta1(self, o: dict) -> typing.Tuple[str | None, dict]:
            obj = DotMap(o)
            obj.apiVersion = "deckhouse.io/v1beta1"
            
            # Merge fields back to hostPort
            obj.spec.hostPort = f"{obj.spec.host}:{obj.spec.port}"
            del obj.spec.host
            del obj.spec.port
            
            return None, obj.toDict()
```

**Handler return type:** `typing.Tuple[str | None, dict]`

| Result        | Return Value               |
| ------------- | -------------------------- |
| ✅ Success<br> | `(None, converted_object)` |
| ❌ Error<br>   | `("error message", {})`    |
<details><summary><b>🎨 Interactive Example</b> (Full flow)</summary>

<a href="https://asciinema.org/a/DHMRw4p49REDkXWS">
  💾 asciinema player
</a>

<img alt="conversion_demo.gif" src="https://github.com/user-attachments/assets/2666109e-5f33-4054-9e95-ee80dca2db3b"/>
</details>

---

### 💡 Good to know about the webhooks
#### 🛠 What's Available in Handlers

| Name             | Description                                       |
| ---------------- | ------------------------------------------------- |
| `DotMap`         | Dict wrapper with dot notation (`obj.spec.field`) |
| `ctx`            | Validation context with request data              |
| `self.snapshots` | Access to Kubernetes snapshots (conversion only)  |
#### ⏱ About webhook registration timing

- Shell-operator reload interval: **5 seconds** (configurable via `SHELL_OPERATOR_RELOAD_INTERVAL_SECONDS` env)
- After creating a webhook CR, wait ~10-15 seconds for it to become active



## Why do we need it, and what problem does it solve?

**Current state:**

- Webhooks are collected from the repository only at Deckhouse compile time
- `webhook-handler` pod reads hook configs and creates subscriptions on startup
- All hooks must exist in the main Deckhouse repo

**Problem:** External modules cannot define their own webhooks — they simply have no way to register them.

**Solution:** Webhook-operator allows defining webhooks dynamically via Kubernetes CRs:

- External modules can create webhooks at runtime
- No need to rebuild Deckhouse
- Webhooks are managed as native Kubernetes resources
- Backward compatible - old implementation continues to work

## Checklist

- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: deckhouse
type: feature
summary: Improved webhook-handler with webhook-operator and CRDs for more complex user flow.
impact_level: default
```

<!---
`impact_level: default` adds to changelog as usual, this is the default that can be omitted
`impact_level: high`    something important for users, the impact will be copied to "Know Before Update" section
`impact_level: low`     omitted in changelog YAML; note there is `type:chore` for chores

Tip for the section field:

  - <kebab-case of a module>, e.g. "cloud-provider-aws", "node-manager"
  - "ci", has forced low impact
  - "docs", includes website changes, should have low impact
  - "candi"
  - "deckhouse-controller"
  - "dhctl"
  - "global-hooks"
  - "go_lib"
  - "helm_lib"
  - "jq_lib"
  - "shell_lib"
  - "testing", has forced low impact
  - "tools", has forced low impact

Find changed sections:

gh pr diff   $PULL_REQUEST_NUMBER   |
  egrep "^([+]{3} b|[-]{3} a)/" |
  cut -d/ -f2- |
  sed 's#^ee/##' |
  sed 's#^fe/##' |
  sed 's#^modules/##' |
  sed 's#[0-9][0-9][0-9]-##' |
  egrep -v 'Makefile' |       # add file exclusion here
  cut -d/ -f1 |
  sort |
  uniq

Find all possible sections (excluding ci):

node -e 'console.log(require("./.github/scripts/js/changelog-find-sections.js")().join("\n"))'
-->

riptide-01 committed:
init operator

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
fix

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
ValidatingWebhook

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
todo

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
template

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
minimums

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
templates

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
struct rework

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
KubernetesContext

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
template as file, enhance tests

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
move templates as package

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
rm unneeded workflows

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
shell-operator prototype as dependency with broken auto-reload

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
run shell-operator as binary with working reload

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
generate conversionWebhook

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
fix tests and spec

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
fix cirrylic

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
finalizer logic

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
manual tests

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
werf go build operator

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
werf go build operator

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
fix ci go generate

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
fix ci go generate

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
fix

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
update tests

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
rm unneded test files

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
update gitignore for dev shell-operator binary

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
ConversionWebhook struct

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
rm trash

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
reconciler init

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
make generate

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
divide validation and conversion

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
resolve some TODOs

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
fix

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
fix copyright

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
change conversion webhook struct

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
conversion reconciler logic

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
fix dmt

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
conversion fix

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
test-files-save

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
fix DMT tests

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
test regen crd

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
struct as in contract, fix template, add tests

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
rm test-files

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
fix

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
documentation update

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
rm e2e

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
fix sample

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

riptide-01 committed:
shell-operator dependency

Signed-off-by: Smyslov Maxim <maksim.smyslov@flant.com>

ldmonster committed:
add copy logic to makefile

Signed-off-by: Pavel Okhlopkov <pavel.okhlopkov@flant.com>

ldmonster committed:
fix ru doc, add echo to makefile

Signed-off-by: Pavel Okhlopkov <pavel.okhlopkov@flant.com>

ldmonster committed:
make it shorter

Signed-off-by: Pavel Okhlopkov <pavel.okhlopkov@flant.com>

ldmonster committed:
bump

Signed-off-by: Pavel Okhlopkov <pavel.okhlopkov@flant.com>

denmaloyreb committed:
docs: Add files for CRD's rendering, fix typos in CRD's

Signed-off-by: Denis.Rebenok <denis.rebenok@flant.ru>

yalosev reviewed (changes requested):
```
context:
  - name: services
    kubernetes:
      group: main
      executeHookOnEvent: []
      executeHookOnSynchronization: false
      keepFullObjectsInMemory: false
      apiVersion: v1
      kind: Service
```

oups, we don't need all these abstractions here
```
      group: main
      executeHookOnEvent: []
      executeHookOnSynchronization: false
      keepFullObjectsInMemory: false
```

z9r5 committed:
Apply suggestions from code review

Co-authored-by: Denis Rebenok <60424108+denmaloyreb@users.noreply.github.com>
Signed-off-by: Artem Kladov <6360800+z9r5@users.noreply.github.com>

z9r5 committed:
Apply suggestions from code review

Co-authored-by: Denis Rebenok <60424108+denmaloyreb@users.noreply.github.com>
Signed-off-by: Artem Kladov <6360800+z9r5@users.noreply.github.com>

Glitchy-Sheep committed:
Make generate for removing old entries in the dependabot.yml

Signed-off-by: Roman Berezkin <roman.berezkin@flant.com>

Glitchy-Sheep committed:
Fix path in werf.inc.yaml

Signed-off-by: Roman Berezkin <roman.berezkin@flant.com>

Glitchy-Sheep committed:
Fix validation webhook to return `(Optional[str], bool)` as `(message, valid)` instead of passing output argument; only context is passed now

- *.tmp files updated to match new signature
- *.yaml files now reflect the new signature
- *.py golden files updated using new template
- validationwebhook_controller_test.go updated to use validation test file instead of conversion one

Signed-off-by: Roman Berezkin <roman.berezkin@flant.com>

Glitchy-Sheep committed:
Regenerate conversion golden files

Signed-off-by: Roman Berezkin <roman.berezkin@flant.com>

Glitchy-Sheep committed:
Fix "forbidden nameSelector" in ValidationWebhook CRD + format doc

The shell-operator doesn't allow nameSelector at all, so we can't use it in the ValidationWebhooks CRD.

- Remove the nameSelector from CRD and conversionwebhook_types
- Regenerate manifests & generate
- Fix go doc comments, so generate looks good in the yaml files

Signed-off-by: Roman Berezkin <roman.berezkin@flant.com>

Glitchy-Sheep committed:
Make shell-operator reload time configurable + less by default (30s -> 5s)

Signed-off-by: Roman Berezkin <roman.berezkin@flant.com>

Glitchy-Sheep committed:
Auto-inject shell-operator service fields in webhook context

Move group, executeHookOnEvent, executeHookOnSynchronization, and keepFullObjectsInMemory from user-specified CRD fields to auto-generated values in hook templates. Users now only configure what to watch, not shell-operator internals.

Signed-off-by: Roman Berezkin <roman.berezkin@flant.com>

Glitchy-Sheep committed:
Fix lint issues

Signed-off-by: Roman Berezkin <roman.berezkin@flant.com>

Glitchy-Sheep committed:
Fix old webhook handler format in sample files

Signed-off-by: Roman Berezkin <roman.berezkin@flant.com>

ldmonster committed:
bump lint

Signed-off-by: Pavel Okhlopkov <pavel.okhlopkov@flant.com>

ldmonster committed:
remove useless

Signed-off-by: Pavel Okhlopkov <pavel.okhlopkov@flant.com>

ldmonster committed:
fix labels

Signed-off-by: Pavel Okhlopkov <pavel.okhlopkov@flant.com>

denmaloyreb commented on modules/002-deckhouse/crds/doc-ru-validationwebhooks.yaml:
```diff
@@ -0,0 +1,245 @@
+          description: |
+            Кастомный ресурс для настройки вебхуков валидации Kubernetes ресурсов.
+
+            Один кастомный ресурс `ValidationWebhook` описывает обработчик валидации для определенных типов ресурсов и операций.
```
```suggestion
            Один кастомный ресурс ValidationWebhook описывает обработчик валидации для определенных типов ресурсов и операций.
```

denmaloyreb commented on modules/002-deckhouse/crds/doc-ru-validationwebhooks.yaml:
```diff
@@ -0,0 +1,245 @@
+                          matchExpressions:
+                            description: |-
+                              Список фильтров на основе выражений.
+                              Итоговый результат — результат пересечения множеств, определяемых всеми фильтрами в списке (операция `И`).
```
```suggestion
                              Список фильтров на основе выражений.
                              
                              Итоговый результат — результат пересечения множеств, определяемых всеми фильтрами в списке (операция `И`).
```

denmaloyreb commented on modules/002-deckhouse/crds/doc-ru-validationwebhooks.yaml:
```diff
@@ -0,0 +1,245 @@
+                                  type: string
+                                operator:
+                                  description: |-
+                                    Представляет соотношение ключа с набором значений.
```
```suggestion
                                    Представляет соотношение ключа с набором значений.
                                    
```

denmaloyreb commented on modules/002-deckhouse/crds/doc-ru-validationwebhooks.yaml:
```diff
@@ -0,0 +1,245 @@
+                      name:
+                        description: |-
+                          Является идентификатором для этого условия соответствия, используемым для стратегического объединения MatchConditions,
+                          а также для предоставления идентификатора для целей ведения журнала. Хорошее имя должно описывать связанное с ним выражение.
```
```suggestion
                         Идентификатор для этого условия соответствия, используемый для стратегического объединения MatchConditions, а также для предоставления идентификатора для целей ведения журнала. Хорошее имя должно описывать связанное с ним выражение.
                         
```

denmaloyreb commented on modules/002-deckhouse/crds/doc-ru-conversionwebhooks.yaml:
```diff
@@ -0,0 +1,157 @@
+                      includeSnapshotsFrom:
+                        description: |
+                          Массив имен привязок kubernetes в хуке.
+                          При указании, список отслеживаемых объектов из этих привязок будет добавлен в контекст привязки в поле `snapshots`. Возможно само-включение.
```
```suggestion
                          Массив имен привязок kubernetes в хуке.
                          
                          Если указано, список отслеживаемых объектов из этих привязок будет добавлен в контекст привязки в поле `snapshots`. Возможно само-включение.
```

denmaloyreb commented on modules/002-deckhouse/crds/doc-ru-conversionwebhooks.yaml:
```diff
@@ -0,0 +1,157 @@
+                  includeSnapshotsFrom:
+                    description: |
+                      Массив имен привязок kubernetes в хуке.
+                      При указании, список отслеживаемых объектов из этих привязок будет добавлен в контекст привязки в поле `snapshots`.
```
```suggestion
                      Массив имен привязок kubernetes в хуке.
                      
                      Если указан, список отслеживаемых объектов из этих привязок будет добавлен в контекст привязки в поле `snapshots`.
```

denmaloyreb commented on modules/002-deckhouse/crds/doc-ru-conversionwebhooks.yaml:
```diff
@@ -0,0 +1,157 @@
+                      jqFilter:
+                        description: |
+                          Необязательный параметр, который указывает фильтрацию событий с использованием синтаксиса jq.
+                          Хук будет запущен на событии "Modified" только если результат фильтра изменился после последнего события.
```
```suggestion
                          Необязательный параметр, который указывает фильтрацию событий с использованием синтаксиса jq.
                          
                          Хук будет запущен на событии «Modified» только если результат фильтра изменился после последнего события.
```

denmaloyreb commented on modules/002-deckhouse/crds/conversionwebhooks.yaml:
```diff
@@ -0,0 +1,277 @@
+                              operator:
+                                description: |-
+                                  operator represents a key's relationship to a set of values.
+                                  Valid operators are In, NotIn, Exists and DoesNotExist.
```
```suggestion
                                  `operator` represents a key's relationship to a set of values.
                                  
                                  Valid operators are `In`, `NotIn`, `Exists` and `DoesNotExist`.
```

denmaloyreb commented on modules/002-deckhouse/crds/conversionwebhooks.yaml:
```diff
@@ -0,0 +1,277 @@
+                                  values is an array of string values. If the operator is In or NotIn,
+                                  the values array must be non-empty. If the operator is Exists or DoesNotExist,
+                                  the values array must be empty. This array is replaced during a strategic
+                                  merge patch.
```
```suggestion
                                  `values` is an array of string values. If the operator is `In` or `NotIn`,
                                  the `values` array must be non-empty. If the operator is `Exists` or `DoesNotExist`,
                                  the `values` array must be empty. This array is replaced during a strategic
                                  merge patch.
```

denmaloyreb commented on modules/002-deckhouse/crds/conversionwebhooks.yaml:
```diff
@@ -0,0 +1,277 @@
+                          description: |-
+                            matchLabels is a map of {key,value} pairs. A single {key,value} in the matchLabels
+                            map is equivalent to an element of matchExpressions, whose key field is "key", the
+                            operator is "In", and the values array contains only "value". The requirements are ANDed.
```
```suggestion
                            `matchLabels` is a map of {key,value} pairs. A single {key,value} in the matchLabels
                            map is equivalent to an element of `matchExpressions`, whose key field is `key`, the
                            operator is `In`, and the values array contains only `value`. The requirements are `ANDed`.
```

denmaloyreb commented on modules/002-deckhouse/crds/conversionwebhooks.yaml:
```diff
@@ -0,0 +1,277 @@
+                                  relates the key and values.
+                                properties:
+                                  key:
+                                    description: key is the label key that the selector
```
```suggestion
                                    description: `key` is the label key that the selector
```

denmaloyreb commented on modules/002-deckhouse/crds/conversionwebhooks.yaml:
```diff
@@ -0,0 +1,277 @@
+                  description: |-
+                    An array of names of kubernetes bindings in a hook.
+                    When specified, a list of monitored objects from
+                    these bindings will be added to the binding context in the snapshots field.
```
```suggestion
                    An array of names of kubernetes bindings in a hook.
                    
                    When specified, a list of monitored objects from
                    these bindings will be added to the binding context in the snapshots field.
```

denmaloyreb commented on modules/002-deckhouse/crds/conversionwebhooks.yaml:
```diff
@@ -0,0 +1,277 @@
+  - name: v1alpha1
+    schema:
+      openAPIV3Schema:
+        description: ConversionWebhook is the Schema for the conversionwebhooks API
```
```suggestion
        description: `ConversionWebhook` is the Schema for the conversionwebhooks API
```

denmaloyreb commented on modules/002-deckhouse/crds/doc-ru-conversionwebhooks.yaml:
```diff
@@ -0,0 +1,157 @@
+                          Тип отслеживаемого ресурса Kubernetes. Это поле обязательно.
+                      labelSelector:
+                        description: |
+                          Селектор меток для фильтрации ресурсов. Результат matchLabels и matchExpressions объединяется через AND. Пустой селектор меток соответствует всем объектам. Нулевой селектор меток не соответствует ни одному объекту.
```
```suggestion
                          Селектор меток для фильтрации ресурсов. Результат `matchLabels` и `matchExpressions` объединяется через `AND`. Пустой селектор меток соответствует всем объектам. Нулевой селектор меток не соответствует ни одному объекту.
```

denmaloyreb commented on modules/002-deckhouse/crds/doc-ru-validationwebhooks.yaml:
```diff
@@ -0,0 +1,245 @@
+                          Тип отслеживаемого ресурса Kubernetes. Это поле обязательно.
+                      labelSelector:
+                        description: |
+                          Селектор меток для фильтрации ресурсов. Результат matchLabels и matchExpressions объединяется через AND. Пустой селектор меток соответствует всем объектам. Нулевой селектор меток не соответствует ни одному объекту.
```
```suggestion
                          Селектор меток для фильтрации ресурсов. Результат `matchLabels` и `matchExpressions` объединяется через `AND`. Пустой селектор меток соответствует всем объектам. Нулевой селектор меток не соответствует ни одному объекту.
```

denmaloyreb commented on modules/002-deckhouse/crds/doc-ru-validationwebhooks.yaml:
```diff
@@ -0,0 +1,245 @@
+                    При указании, список отслеживаемых объектов из этих привязок будет добавлен в контекст привязки в поле snapshots. Возможно само-включение.
+                labelSelector:
+                  description: |
+                    Селектор меток для фильтрации ресурсов. Результат matchLabels и matchExpressions объединяется через AND. Пустой селектор меток соответствует всем объектам. Нулевой селектор меток не соответствует ни одному объекту.
```
```suggestion
                    Селектор меток для фильтрации ресурсов. Результат `matchLabels` и `matchExpressions` объединяется через `AND`. Пустой селектор меток соответствует всем объектам. Нулевой селектор меток не соответствует ни одному объекту.
```

denmaloyreb commented on modules/002-deckhouse/crds/validationwebhooks.yaml:
```diff
@@ -0,0 +1,500 @@
+                                  operator:
+                                    description: |-
+                                      operator represents a key's relationship to a set of values.
+                                      Valid operators are In, NotIn, Exists and DoesNotExist.
```
```suggestion
                                    description: `key` is the label key that the selector
                                      applies to.
                                    type: string
                                  operator:
                                    description: |-
                                      `operator` represents a key's relationship to a set of values.
                                      Valid operators are `In`, `NotIn`, `Exists` and `DoesNotExist`.
```

denmaloyreb commented on modules/002-deckhouse/crds/validationwebhooks.yaml:
```diff
@@ -0,0 +1,500 @@
+                                      values is an array of string values. If the operator is In or NotIn,
+                                      the values array must be non-empty. If the operator is Exists or DoesNotExist,
+                                      the values array must be empty. This array is replaced during a strategic
+                                      merge patch.
```
```suggestion
                                      `values` is an array of string values. If the operator is `In` or `NotIn`,
                                      the values array must be non-empty. If the operator is `Exists` or `DoesNotExist`,
                                      the values array must be empty. This array is replaced during a strategic
                                      merge patch.
```

denmaloyreb commented on modules/002-deckhouse/crds/validationwebhooks.yaml:
```diff
@@ -0,0 +1,500 @@
+                  label selector matches no objects.
+                properties:
+                  matchExpressions:
+                    description: matchExpressions is a list of label selector requirements.
```
```suggestion
                    description: `matchExpressions` is a list of label selector requirements.
```

denmaloyreb commented on modules/002-deckhouse/crds/validationwebhooks.yaml:
```diff
@@ -0,0 +1,500 @@
+                        'object' - The object from the incoming request. The value is null for DELETE requests.
+                        'oldObject' - The existing object. The value is null for CREATE requests.
+                        'request' - Attributes of the admission request(/pkg/apis/admission/types.go#AdmissionRequest).
+                        'authorizer' - A CEL Authorizer. May be used to perform authorization checks for the principal (user or service account) of the request.
```
```suggestion
                        'object': The object from the incoming request. The value is null for `DELETE` requests.
                        'oldObject': The existing object. The value is null for `CREATE` requests.
                        'request': Attributes of the admission request(/pkg/apis/admission/types.go#AdmissionRequest).
                        'authorizer': A CEL Authorizer. May be used to perform authorization checks for the principal (user or service account) of the request.
```

Glitchy-Sheep committed:
Public/Private fields refactor + better general code structure

- Make struct fields private, add constructor functions
- Extract path constants and add helper methods (webhookDir, webhookFilePath)
- Replace string concatenation with filepath.Join
- Improve logging with structured slog attributes
- Update main.go and tests to use new constructors

Signed-off-by: Roman Berezkin <roman.berezkin@flant.com>

ldmonster committed:
Update modules/002-deckhouse/crds/doc-ru-validationwebhooks.yaml

Co-authored-by: Denis Rebenok <60424108+denmaloyreb@users.noreply.github.com>
Signed-off-by: Pavel Okhlopkov <36456348+ldmonster@users.noreply.github.com>

ldmonster committed:
Update modules/002-deckhouse/crds/doc-ru-validationwebhooks.yaml

Co-authored-by: Denis Rebenok <60424108+denmaloyreb@users.noreply.github.com>
Signed-off-by: Pavel Okhlopkov <36456348+ldmonster@users.noreply.github.com>

ldmonster committed:
Update modules/002-deckhouse/crds/doc-ru-validationwebhooks.yaml

Co-authored-by: Denis Rebenok <60424108+denmaloyreb@users.noreply.github.com>
Signed-off-by: Pavel Okhlopkov <36456348+ldmonster@users.noreply.github.com>

ldmonster committed:
Update modules/002-deckhouse/crds/doc-ru-validationwebhooks.yaml

Co-authored-by: Denis Rebenok <60424108+denmaloyreb@users.noreply.github.com>
Signed-off-by: Pavel Okhlopkov <36456348+ldmonster@users.noreply.github.com>

ldmonster committed:
Update modules/002-deckhouse/crds/doc-ru-validationwebhooks.yaml

Co-authored-by: Denis Rebenok <60424108+denmaloyreb@users.noreply.github.com>
Signed-off-by: Pavel Okhlopkov <36456348+ldmonster@users.noreply.github.com>

ldmonster committed:
Update modules/002-deckhouse/crds/validationwebhooks.yaml

Co-authored-by: Denis Rebenok <60424108+denmaloyreb@users.noreply.github.com>
Signed-off-by: Pavel Okhlopkov <36456348+ldmonster@users.noreply.github.com>

ldmonster committed:
Update modules/002-deckhouse/crds/validationwebhooks.yaml

Co-authored-by: Denis Rebenok <60424108+denmaloyreb@users.noreply.github.com>
Signed-off-by: Pavel Okhlopkov <36456348+ldmonster@users.noreply.github.com>

ldmonster committed:
Update modules/002-deckhouse/crds/validationwebhooks.yaml

Co-authored-by: Denis Rebenok <60424108+denmaloyreb@users.noreply.github.com>
Signed-off-by: Pavel Okhlopkov <36456348+ldmonster@users.noreply.github.com>

ldmonster committed:
Update modules/002-deckhouse/crds/validationwebhooks.yaml

Co-authored-by: Denis Rebenok <60424108+denmaloyreb@users.noreply.github.com>
Signed-off-by: Pavel Okhlopkov <36456348+ldmonster@users.noreply.github.com>

ldmonster committed:
Update modules/002-deckhouse/crds/validationwebhooks.yaml

Co-authored-by: Denis Rebenok <60424108+denmaloyreb@users.noreply.github.com>
Signed-off-by: Pavel Okhlopkov <36456348+ldmonster@users.noreply.github.com>

Glitchy-Sheep committed:
Remove unused fields from KubernetesContext

- Remove ResynchronizationPeriod and IncludeSnapshotsFrom fields
- These fields have no effect since context bindings use:
    - executeHookOnEvent: []
    - executeHookOnSynchronization: false
- Regenerate deepcopy and CRD manifests

Signed-off-by: Roman Berezkin <roman.berezkin@flant.com>

z9r5 commented on modules/002-deckhouse/crds/validationwebhooks.yaml:
```diff
@@ -0,0 +1,500 @@
+                        Name must be a qualified name consisting of alphanumeric characters, '-', '_' or '.', and
+                        must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or
+                        '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]') with an
+                        optional DNS subdomain prefix and '/' (e.g. 'example.com/MyName')
```
Seems ok

z9r5 committed:
Apply suggestions from code review

Co-authored-by: Denis Rebenok <60424108+denmaloyreb@users.noreply.github.com>
Signed-off-by: Artem Kladov <6360800+z9r5@users.noreply.github.com>

z9r5 committed:
Apply suggestions from code review

Co-authored-by: Denis Rebenok <60424108+denmaloyreb@users.noreply.github.com>
Signed-off-by: Artem Kladov <6360800+z9r5@users.noreply.github.com>

Glitchy-Sheep committed:
make generate manifests (re-generate manifests files)

Signed-off-by: Roman Berezkin <roman.berezkin@flant.com>

Glitchy-Sheep committed:
Use |- block scalar pattern to fix yaml parser errors

Signed-off-by: Roman Berezkin <roman.berezkin@flant.com>

Glitchy-Sheep committed:
Remove unnecessary ctrl.Result.Requeue = true triggers on error paths

- Those will be handled automatically by the controller-runtime

Signed-off-by: Roman Berezkin <roman.berezkin@flant.com>

Glitchy-Sheep committed:
Fix documentation typo (kubernetes -> Kubernetes)

Doc-comments were changed to keep the documentation (make manifests) in a desired format

Signed-off-by: Roman Berezkin <roman.berezkin@flant.com>

Glitchy-Sheep committed:
Restore IncludeSnapshotsFrom field

Signed-off-by: Roman Berezkin <roman.berezkin@flant.com>

Glitchy-Sheep committed:
Migrate deckhouse validating webhooks to ValidationWebhook CRDs

Signed-off-by: Roman Berezkin <roman.berezkin@flant.com>

Glitchy-Sheep committed:
Fix jqFilter (no new lines)

Signed-off-by: Roman Berezkin <roman.berezkin@flant.com>

ldmonster committed:
Update modules/002-deckhouse/crds/doc-ru-validationwebhooks.yaml

Co-authored-by: Artem Kladov <6360800+z9r5@users.noreply.github.com>
Signed-off-by: Pavel Okhlopkov <36456348+ldmonster@users.noreply.github.com>

Glitchy-Sheep committed:
Autogenerate includeSnapshotsFrom based on kubernetes context entries

Signed-off-by: Roman Berezkin <roman.berezkin@flant.com>

Glitchy-Sheep committed:
Remove redundant includeSnapshotsFrom sections from template and docs

Signed-off-by: Roman Berezkin <roman.berezkin@flant.com>

Glitchy-Sheep committed:
Add CRD cleanup finalizer to prevent stale webhook conversion on ConversionWebhook deletion

Signed-off-by: Roman Berezkin <roman.berezkin@flant.com>

Glitchy-Sheep committed:
Enable old hooks to support both types of webhooks

Signed-off-by: Roman Berezkin <roman.berezkin@flant.com>

ldmonster committed:
Merge commit 'deb61e9911b0d9f26ab941718016bd9d06a1b534' into feature/webhook-operator

ldmonster committed:
Merge commit 'c36e9afb1a59d4dcd823670344245d33ec926882' into feature/webhook-operator
